src/OneSpanSign.ts

```diff
@@ -25,55 +20,40 @@ export class OneSpanSign {
    *
    * - A list of server URLs can be found at {@link https://community.onespan.com/documentation/onespan-sign/guides/quick-start-guides/developer/environment-urls-ip-addresses | Environment URLs & IP Addresses (OneSpan)}.
    */
-  constructor(private readonly apiKey: string, private readonly apiUrl: string) {}
+  constructor(apiKey: string, apiUrl: string) {
+    this.api = new Api(apiKey, apiUrl);
+  }
 
   /**
-   * Creates a package (transaction).
-   *
-   * @param payload - Package initial information
-   * @returns A payload that contains the ID of the newly created package
+   * Individual resource objects are lazily initialized. When a getter for the resource is called,
+   * the resource object will be initialized and shadows the getter function as an object property.
    *
-   * @remarks
-   * - {@link https://community.onespan.com/products/onespan-sign/sandbox#/Packages/api.packages.post | REST API documentation (OneSpan)}
-   *
-   * - {@link https://community.onespan.com/documentation/onespan-sign/guides/feature-guides/developer/creating-transaction-sender | Creating a Transaction for a Sender (OneSpan)}
+   * Inspiration: https://stackoverflow.com/a/37978698/14163928
    */
-  public async createPackage(payload: CreatePackageRequestPayload): Promise<CreatePackageResponsePayload> {
-    const response = await Api.post(`${this.apiUrl}/api/packages`)
-      .withAuthorizationHeader(`Basic ${this.apiKey}`)
-      .withBody(payload)
-      .fetch();
 
-    return (await response.json()) as CreatePackageResponsePayload;
+  /**
+   * Document resource
+   */
+  public get documents(): DocumentResource {
+    const documents = new DocumentResource(this.api);
+    Object.defineProperty(this, 'documents', {
```

Comment:
Do you need to do this to be able to have this 'lazy loading' of resources? 

This looks a bit contrived imo. What is the advantage of this pattern?

Can you point to documentation reegarding this pattern?

Reply:
This pattern is inspired by this [Stackoverflow thread](https://stackoverflow.com/a/37978698/14163928).

When we create a new instance of this class, we are essentially creating a new JS object with the prototype pointing to the `OneSpanSign` object:
```ts
const client = new OneSpanSign(API_KEY, API_URL);

// is the equivalent of:

const client = Object.create(OneSpanSign);
```

`documents` is not actually defined on `client` but rather on `OneSpanSign`. When we call `client.documents`, we are calling this getter function from its prototype (based on JS prototype inheritance).

```ts
console.log(client);

// Prints:
// client = {
//   api: apiInstance
// }

// client.document -> client.prototype.document (OneSpanSign.document)
```

To memoize the `DocumentResource` instance once we initialize it, we define a `document` property directly on the `client` object to shadow the call to the prototype's getter function:

```ts
console.log(client);

// Prints:
// client = {
//   api: apiInstance
// }

Object.defineProperty(client, 'documents', documentResource);

console.log(client);

// Prints:
// client = {
//   api: apiInstance,
//   document: documentResource
// }

// client.document -> documentResource;

// client.prototype.document -> OneSpanSign.document
```

Hope this makes sense!

Reply:
Yep, that helps a lot, ty.

This is pretty cool, but you know, I have to be that guy and ask: do we need to memoize/lazy load the resources?

Reply:
Right now we only have two resource classes so it probably won't make much difference. But since we aim to support all APIs of OneSpan Sign, we could get as many as 10+ resource classes in the end. I find it unlikely that a user needs to use all resources in one go. So I added this pattern to save the trouble of initializing everything.

Then again, it may not be expensive to initialize everything since we are not storing huge amounts of data. Plus, this library is designed to run server-side.

Reply:
Nah, leave it like this for now. I think it's cool to give it a shot--I just wanted to make sure we are doing this for a reason and not over-engineering this.